File: app/src/main/java/github/leavesc/wifip2p/ReceiveFileActivity.java
package github.leavesc.wifip2p;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.net.wifi.p2p.WifiP2pDevice;
import android.net.wifi.p2p.WifiP2pInfo;
import android.net.wifi.p2p.WifiP2pManager;
import android.os.Bundle;
import android.os.IBinder;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.core.app.ActivityCompat;

import com.bumptech.glide.Glide;

import java.io.File;
import java.util.Collection;

import github.leavesc.wifip2p.broadcast.DirectBroadcastReceiver;
import github.leavesc.wifip2p.callback.DirectActionListener;
import github.leavesc.wifip2p.model.FileTransfer;
import github.leavesc.wifip2p.service.WifiServerService;

/**
 * @Author: leavesC
 * @Date: 2019/2/27 23:52
 * @Desc:
 * @Github：https://github.com/leavesC
 */
public class ReceiveFileActivity extends BaseActivity {

    private ImageView iv_image;

    private TextView tv_log;

    private ProgressDialog progressDialog;

    private WifiP2pManager wifiP2pManager;

    private WifiP2pManager.Channel channel;

    private boolean connectionInfoAvailable;

    private BroadcastReceiver broadcastReceiver;

    private WifiServerService wifiServerService;

    private final ServiceConnection serviceConnection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            WifiServerService.WifiServerBinder binder = (WifiServerService.WifiServerBinder) service;
            wifiServerService = binder.getService();
            wifiServerService.setProgressChangListener(progressChangListener);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            if (wifiServerService != null) {
                wifiServerService.setProgressChangListener(null);
                wifiServerService = null;
            }
            bindService();
        }
    };

    private final DirectActionListener directActionListener = new DirectActionListener() {
        @Override
        public void wifiP2pEnabled(boolean enabled) {
            log("wifiP2pEnabled: " + enabled);
        }

        @Override
        public void onConnectionInfoAvailable(WifiP2pInfo wifiP2pInfo) {
            log("onConnectionInfoAvailable");
            log("isGroupOwner：" + wifiP2pInfo.isGroupOwner);
            log("groupFormed：" + wifiP2pInfo.groupFormed);
            if (wifiP2pInfo.groupFormed && wifiP2pInfo.isGroupOwner) {
                connectionInfoAvailable = true;
                if (wifiServerService != null) {
                    startService(WifiServerService.class);
                }
            }
        }

        @Override
        public void onDisconnection() {
            connectionInfoAvailable = false;
            log("onDisconnection");
        }

        @Override
        public void onSelfDeviceAvailable(WifiP2pDevice wifiP2pDevice) {
            log("onSelfDeviceAvailable");
            log(wifiP2pDevice.toString());
        }

        @Override
        public void onPeersAvailable(Collection<WifiP2pDevice> wifiP2pDeviceList) {
            log("onPeersAvailable,size:" + wifiP2pDeviceList.size());
            for (WifiP2pDevice wifiP2pDevice : wifiP2pDeviceList) {
                log(wifiP2pDevice.toString());
            }
        }

        @Override
        public void onChannelDisconnected() {
            log("onChannelDisconnected");
        }
    };

    private final WifiServerService.OnProgressChangListener progressChangListener = new WifiServerService.OnProgressChangListener() {
        @Override
        public void onProgressChanged(final FileTransfer fileTransfer, final int progress) {
            runOnUiThread(() -> {
                progressDialog.setMessage("文件名： " + fileTransfer.getFileName());
                progressDialog.setProgress(progress);
                progressDialog.show();
            });
        }

        @Override
        public void onTransferFinished(final File file) {
            runOnUiThread(() -> {
                progressDialog.cancel();
                if (file != null && file.exists()) {
                    Glide.with(ReceiveFileActivity.this).load(file.getPath()).into(iv_image);
                }
            });
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_receive_file);
        initView();
        wifiP2pManager = (WifiP2pManager) getSystemService(WIFI_P2P_SERVICE);
        if (wifiP2pManager == null) {
            finish();
            return;
        }
        channel = wifiP2pManager.initialize(this, getMainLooper(), directActionListener);
        broadcastReceiver = new DirectBroadcastReceiver(wifiP2pManager, channel, directActionListener);
        registerReceiver(broadcastReceiver, DirectBroadcastReceiver.getIntentFilter());
        bindService();
    }

    private void initView() {
        setTitle("接收文件");
        iv_image = findViewById(R.id.iv_image);
        tv_log = findViewById(R.id.tv_log);
        findViewById(R.id.btnCreateGroup).setOnClickListener(v -> {
            if (ActivityCompat.checkSelfPermission(ReceiveFileActivity.this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                return;
            }
            wifiP2pManager.createGroup(channel, new WifiP2pManager.ActionListener() {
                @Override
                public void onSuccess() {
                    log("createGroup onSuccess");
                    dismissLoadingDialog();
                    showToast("onSuccess");
                }

                @Override
                public void onFailure(int reason) {
                    log("createGroup onFailure: " + reason);
                    dismissLoadingDialog();
                    showToast("onFailure");
                }
            });
        });
        findViewById(R.id.btnRemoveGroup).setOnClickListener(v -> removeGroup());
        progressDialog = new ProgressDialog(this);
        progressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
        progressDialog.setCancelable(false);
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.setTitle("正在接收文件");
        progressDialog.setMax(100);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (wifiServerService != null) {
            wifiServerService.setProgressChangListener(null);
            unbindService(serviceConnection);
        }
        unregisterReceiver(broadcastReceiver);
        stopService(new Intent(this, WifiServerService.class));
        if (connectionInfoAvailable) {
            removeGroup();
        }
    }

    private void removeGroup() {
        wifiP2pManager.removeGroup(channel, new WifiP2pManager.ActionListener() {
            @Override
            public void onSuccess() {
                log("removeGroup onSuccess");
                showToast("onSuccess");
            }

            @Override
            public void onFailure(int reason) {
                log("removeGroup onFailure");
                showToast("onFailure");
            }
        });
    }

    private void log(String log) {
        tv_log.append(log + "\n");
        tv_log.append("----------" + "\n");
    }

    private void bindService() {
        Intent intent = new Intent(ReceiveFileActivity.this, WifiServerService.class);
        bindService(intent, serviceConnection, BIND_AUTO_CREATE);
    }

}

File: app/src/main/java/github/leavesczy/wifip2p/ReceiveFileActivity.kt
package github.leavesczy.wifip2p

import android.Manifest
import android.app.ProgressDialog
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.net.wifi.p2p.WifiP2pDevice
import android.net.wifi.p2p.WifiP2pInfo
import android.net.wifi.p2p.WifiP2pManager
import android.os.Bundle
import android.os.IBinder
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.app.ActivityCompat
import com.bumptech.glide.Glide
import github.leavesczy.wifip2p.service.DirectActionListener
import github.leavesczy.wifip2p.service.DirectBroadcastReceiver
import github.leavesczy.wifip2p.service.FileTransfer
import github.leavesczy.wifip2p.service.WifiServerService
import github.leavesczy.wifip2p.service.WifiServerService.OnProgressChangListener
import github.leavesczy.wifip2p.service.WifiServerService.WifiServerBinder
import java.io.File

/**
 * @Author: leavesCZY
 * @Desc:
 */
class ReceiveFileActivity : BaseActivity() {

    private lateinit var ivImage: ImageView

    private lateinit var tvLog: TextView

    private lateinit var progressDialog: ProgressDialog

    private val progressChangListener: OnProgressChangListener = object : OnProgressChangListener {
        override fun onProgressChanged(fileTransfer: FileTransfer, progress: Int) {
            runOnUiThread {
                progressDialog.setMessage("文件名： " + fileTransfer.fileName)
                progressDialog.progress = progress
                progressDialog.show()
            }
        }

        override fun onTransferFinished(file: File?) {
            runOnUiThread {
                progressDialog.cancel()
                if (file != null && file.exists()) {
                    Glide.with(this@ReceiveFileActivity).load(file.path).into(ivImage)
                }
            }
        }
    }

    private lateinit var wifiP2pManager: WifiP2pManager

    private var channel: WifiP2pManager.Channel? = null

    private var connectionInfoAvailable = false

    private var broadcastReceiver: BroadcastReceiver? = null

    private var wifiServerService: WifiServerService? = null

    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            val binder = service as WifiServerBinder
            wifiServerService = binder.service
            wifiServerService!!.setProgressChangListener(progressChangListener)
        }

        override fun onServiceDisconnected(name: ComponentName) {
            if (wifiServerService != null) {
                wifiServerService!!.setProgressChangListener(null)
                wifiServerService = null
            }
            bindService()
        }
    }
    private val directActionListener: DirectActionListener = object : DirectActionListener {
        override fun wifiP2pEnabled(enabled: Boolean) {
            log("wifiP2pEnabled: $enabled")
        }

        override fun onConnectionInfoAvailable(wifiP2pInfo: WifiP2pInfo) {
            log("onConnectionInfoAvailable")
            log("isGroupOwner：" + wifiP2pInfo.isGroupOwner)
            log("groupFormed：" + wifiP2pInfo.groupFormed)
            if (wifiP2pInfo.groupFormed && wifiP2pInfo.isGroupOwner) {
                connectionInfoAvailable = true
                if (wifiServerService != null) {
                    startService(WifiServerService::class.java)
                }
            }
        }

        override fun onDisconnection() {
            connectionInfoAvailable = false
            log("onDisconnection")
        }

        override fun onSelfDeviceAvailable(wifiP2pDevice: WifiP2pDevice?) {
            log("onSelfDeviceAvailable")
            log(wifiP2pDevice.toString())
        }

        override fun onPeersAvailable(wifiP2pDeviceList: Collection<WifiP2pDevice>) {
            log("onPeersAvailable,size:" + wifiP2pDeviceList.size)
            for (wifiP2pDevice in wifiP2pDeviceList) {
                log(wifiP2pDevice.toString())
            }
        }

        override fun onChannelDisconnected() {
            log("onChannelDisconnected")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_receive_file)
        initView()
        val mWifiP2pManager = getSystemService(WIFI_P2P_SERVICE) as? WifiP2pManager
        if (mWifiP2pManager == null) {
            finish()
            return
        }
        wifiP2pManager = mWifiP2pManager
        channel = wifiP2pManager.initialize(this, mainLooper, directActionListener)
        broadcastReceiver = DirectBroadcastReceiver(wifiP2pManager, channel, directActionListener)
        registerReceiver(broadcastReceiver, DirectBroadcastReceiver.intentFilter)
        bindService()
    }

    private fun initView() {
        setTitle("接收文件")
        ivImage = findViewById(R.id.ivImage)
        tvLog = findViewById(R.id.tvLog)
        findViewById<View>(R.id.btnCreateGroup).setOnClickListener {
            if (ActivityCompat.checkSelfPermission(
                    this@ReceiveFileActivity,
                    Manifest.permission.ACCESS_FINE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                return@setOnClickListener
            }
            wifiP2pManager.createGroup(channel, object : WifiP2pManager.ActionListener {
                override fun onSuccess() {
                    log("createGroup onSuccess")
                    dismissLoadingDialog()
                    showToast("onSuccess")
                }

                override fun onFailure(reason: Int) {
                    log("createGroup onFailure: $reason")
                    dismissLoadingDialog()
                    showToast("onFailure")
                }
            })
        }
        findViewById<View>(R.id.btnRemoveGroup).setOnClickListener {
            removeGroup()
        }
        progressDialog = ProgressDialog(this).apply {
            setProgressStyle(ProgressDialog.STYLE_HORIZONTAL)
            setCancelable(false)
            setCanceledOnTouchOutside(false)
            max = 100
            setTitle("正在接收文件")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (wifiServerService != null) {
            wifiServerService?.setProgressChangListener(null)
            unbindService(serviceConnection)
        }
        if (broadcastReceiver != null) {
            unregisterReceiver(broadcastReceiver)
        }
        stopService(Intent(this, WifiServerService::class.java))
        if (connectionInfoAvailable) {
            removeGroup()
        }
    }

    private fun removeGroup() {
        wifiP2pManager.removeGroup(channel, object : WifiP2pManager.ActionListener {
            override fun onSuccess() {
                log("removeGroup onSuccess")
                showToast("onSuccess")
            }

            override fun onFailure(reason: Int) {
                log("removeGroup onFailure")
                showToast("onFailure")
            }
        })
    }

    private fun log(log: String) {
        tvLog.append(log)
        tvLog.append("----------")
    }

    private fun bindService() {
        val intent = Intent(this@ReceiveFileActivity, WifiServerService::class.java)
        bindService(intent, serviceConnection, BIND_AUTO_CREATE)
    }
}